src/main/java/si/mazi/rescu/clients/ApacheConnection.java

```diff
@@ -0,0 +1,217 @@
+    
+    private static final Logger log = LoggerFactory.getLogger(ApacheConnection.class);
+    
+    private final String url;
+    private final Proxy proxy;
+    private final HttpClientBuilder builder = HttpClients.custom();
+    private final RequestConfig.Builder requestConfig = RequestConfig.custom();
+            
+    private boolean executed = false;
+    private HttpMethod method = HttpMethod.GET;
+    private ByteArrayOutputStream out;
+    private final Map<String, String> headers = new HashMap<>();
+    
+    private CloseableHttpResponse response;
+
+    private ApacheConnection(String url, Proxy proxy) {
+        super();
+        this.url = url;
+        this.proxy = proxy;
+    }
+    
+    public static HttpConnection create(String url, Proxy proxy) throws MalformedURLException, IOException {
+        return new ApacheConnection(url, proxy);
+    }
+    
+    private void exec() throws ClientProtocolException, IOException {
+        if (executed) {
+            return;
+        }
+        HttpClientBuilder clientBuilder = builder
+                .setDefaultRequestConfig(requestConfig.build());
+        if (proxy != null && proxy != Proxy.NO_PROXY) {
+            InetSocketAddress address = (InetSocketAddress) proxy.address(); 
+            HttpHost proxy = new HttpHost(address.getHostName(), address.getPort());
+            clientBuilder.setProxy(proxy);
+        }
+        try ( CloseableHttpClient client = clientBuilder.build() ) {
+        
+            HttpRequestBase request = createRequest(method, url);
+            
+            if (request instanceof HttpEntityEnclosingRequestBase && out != null) {
+                HttpEntityEnclosingRequestBase req = (HttpEntityEnclosingRequestBase) request;
+                HttpEntity entity = req.getEntity();
+                if (entity == null) {
+                    entity = new ByteArrayEntity(out.toByteArray());
+                    req.setEntity(entity);
+                }
+            }
+            headers.forEach(request::addHeader);
+            response = client.execute(request);
+            executed = true;
+        }
+    }
+
+    private static HttpRequestBase createRequest(HttpMethod method, String url) {
+        switch(method) {
+            case GET: return new HttpGet(url); 
+            case POST: return new HttpPost(url);
+            case PUT: return new HttpPut(url);
+            case DELETE: return new HttpDelete(url);
+            case HEAD: return new HttpHead(url); 
+            case OPTIONS: return new HttpOptions(url);
+            case PATCH: return new HttpPatch(url);
+            default: throw new RuntimeException("Not supported http method: " + method);
+        }
+    }
+
+    @Override
+    public String getHeaderField(String name) {
+        if (!executed) {
+            throw new RuntimeException("Request is not executed yet.");
+        }
+        return response.getFirstHeader(name).getValue();
```

Comment:
NPE here, if the header with the name is absent.

Reply:
will change, good catch